### ISSUE TITLE
partial instanced file failed OTS check

### ISSUE THREAD

**Issue by qxliu76** [fonttools/fonttools#3025]
I used this command to generate a partial instanced file: `fonttools varLib.instancer --no-overlap-flag --no-recalc-timestamp catamaran.ttf wght=100:700`, and the generated file failed OTS check, see:
```
ERROR at ../src/variations.cc:244 (ParseVariationData)
ERROR: Variations: Tuple coordinate not in the range [-1.0, 1.0]: 1.66254
ERROR at ../src/gvar.cc:64 (ParseGlyphVariationDataArray)
ERROR: gvar: Failed to parse GlyphVariationData
ERROR: gvar: Failed to read glyph variation data array
ERROR: Dropping all Variation tables
ERROR: avar: Discarding Variations table
ERROR: Table discarded
ERROR: fvar: Discarding Variations table
ERROR: Table discarded
ERROR: gvar: Discarding Variations table
ERROR: Table discarded
ERROR at ../src/variations.cc:33 (ParseVariationRegionList)
ERROR: Variations: Required fvar table is missing
ERROR at ../src/variations.cc:136 (ParseItemVariationStore)
ERROR: Variations: Failed to parse variation region list
ERROR: HVAR: Failed to parse item variation store
ERROR: Dropping all Variation tables
ERROR: HVAR: Discarding Variations table
ERROR: Table discarded
ERROR: avar: Discarding Variations table
ERROR: Table discarded
ERROR: fvar: Discarding Variations table
ERROR: Table discarded
ERROR: gvar: Discarding Variations table
ERROR: Table discarded
ERROR at ../src/variations.cc:33 (ParseVariationRegionList)
ERROR: Variations: Required fvar table is missing
ERROR at ../src/variations.cc:136 (ParseItemVariationStore)
ERROR: Variations: Failed to parse variation region list
ERROR: GDEF: Invalid item variation store
ERROR at ../src/ots.cc:726 (ProcessGeneric)
ERROR: GDEF: Failed to parse table
Failed to sanitize file!
```
[catamaran.zip](https://github.com/fonttools/fonttools/files/10913188/catamaran.zip)

**Comment by behdad** [comment]
> ERROR: Variations: Tuple coordinate not in the range [-1.0, 1.0]: 1.66254

This is ots being annoyingly too conservating.

**Comment by behdad** [comment]
cc @khaledhosny

**Comment by khaledhosny** [comment]
I believe that variation code was written by @jfkthame

**Comment by jfkthame** [comment]
Yes, I think I'm the guilty party, though my memory of it is a bit vague at this point.
Is it actually meaningful/valid to have tuple coords outside the [-1.0, 1.0] range? I thought these were all in the normalized coordinate space where min is mapped to -1.0 and max to +1.0, so it seems unexpected to refer to anything beyond that.

**Comment by behdad** [comment]
> Is it actually meaningful/valid to have tuple coords outside the [-1.0, 1.0] range? I thought these were all in the normalized coordinate space where min is mapped to -1.0 and max to +1.0, so it seems unexpected to refer to anything beyond that.

When partially-instantiating fonts (changing the axis range limits) we sometimes create tuples with those characteristics, since the format allows for numbers in [-2.0, 2.0) range.

**Comment by behdad** [comment]
This OTS patch makes it pass:
```patch
diff --git a/src/variations.cc b/src/variations.cc
index 55afd97..acc32dc 100644
--- a/src/variations.cc
+++ b/src/variations.cc
@@ -47,9 +47,6 @@ bool ParseVariationRegionList(const ots::Font* font, const uint8_t* data, const
       if (startCoord > peakCoord || peakCoord > endCoord) {
         return OTS_FAILURE_MSG("Region axis coordinates out of order");
       }
-      if (startCoord < -0x4000 || endCoord > 0x4000) {
-        return OTS_FAILURE_MSG("Region axis coordinate out of range");
-      }
       if ((peakCoord < 0 && endCoord > 0) ||
           (peakCoord > 0 && startCoord < 0)) {
         return OTS_FAILURE_MSG("Invalid region axis coordinates");
@@ -215,9 +212,6 @@ bool ParseVariationData(const Font* font, const uint8_t* data, size_t length,
         if (!subtable.ReadS16(&coordinate)) {
           return OTS_FAILURE_MSG("Failed to read tuple coordinate");
         }
-        if (coordinate < -0x4000 || coordinate > 0x4000) {
-          return OTS_FAILURE_MSG("Tuple coordinate not in the range [-1.0, 1.0]: %g", coordinate / 16384.);
-        }
       }
     }
 
@@ -228,9 +222,6 @@ bool ParseVariationData(const Font* font, const uint8_t* data, size_t length,
         if (!subtable.ReadS16(&coordinate)) {
           return OTS_FAILURE_MSG("Failed to read tuple coordinate");
         }
-        if (coordinate < -0x4000 || coordinate > 0x4000) {
-          return OTS_FAILURE_MSG("Tuple coordinate not in the range [-1.0, 1.0]: %g", coordinate / 16384.);
-        }
         startTuple.push_back(coordinate);
       }
 
@@ -240,9 +231,6 @@ bool ParseVariationData(const Font* font, const uint8_t* data, size_t length,
         if (!subtable.ReadS16(&coordinate)) {
           return OTS_FAILURE_MSG("Failed to read tuple coordinate");
         }
-        if (coordinate < -0x4000 || coordinate > 0x4000) {
-          return OTS_FAILURE_MSG("Tuple coordinate not in the range [-1.0, 1.0]: %g", coordinate / 16384.);
-        }
         endTuple.push_back(coordinate);
       }
 
```

**Comment by behdad** [comment]
The font is available here:
https://fonts.google.com/specimen/Catamaran

**Comment by anthrotype** [comment]
That's unfortunate. If the instancer had to comply with that, it'd have to create additional tuples which, increasing file size.
So +1 to patching OTS as per behdad's patch.

Then I believe we'd also need to relax the spec itself about this, as it currently says stuff like:

https://learn.microsoft.com/en-us/typography/opentype/spec/otvarcommonformats#variation-regions

> {start/peak/endCoords} must all be within the range -1.0 to +1.0.

**Comment by anthrotype** [comment]
> normalized coordinate space where min is mapped to -1.0 and max to +1.0

yes, that relationship would still continue to hold, even we we set 1.0 < max < 2.0 (or -2.0 <= min <  -1.0), and the clamping to -1.0/+1.0 still occurs and the part of "tents" beyond the -1.0/+1.0 limits is never evaluated; what this allows us is being able to restrict a tent's min or max without having to add an additional tent, whenever ...

Let's see if I can sketch this (I'm not great drawing with a mouse).

Say we have a tent with (start, peak, end) coordinates equal to (0.0, 0.7, 1.0) painted black in the picture, and the user requests to restrict end coord to 0.8; after dividing the tent's coords by 0.8 we obtain (0.0, 0.875, 1.25); since we can encode 1.25 in our F2Dot14, we can just keep this single tent with the rebased coordinates.

![image](https://user-images.githubusercontent.com/6939968/223700748-f405802b-17b4-470c-99f7-fe7932fbc9bc.png)

If we had to restrict ourselves to start/end not exceeding -1.0/+1.0, we need _two_ triangles (red and green in the second picture below) in order to obtain the same effect of the single initial black triangle after the latter has been restricted to end=0.8 (in the old coordinate system, 1.0 in the new).

![image](https://user-images.githubusercontent.com/6939968/223700781-0160e850-5433-4ad9-8cb5-f2bb4623bd05.png)

@jfkthame Does that make sense to you?

**Comment by jfkthame** [comment]
Yes, that makes sense.... but I note that it doesn't cover all cases. The instancer will still need to be able to create additional tuples / tents in cases that don't fit the constraint here.

E.g. with your example, if the original peak were instead at 0.2 rather than 0.8, and the user requests to restrict the end coord to 0.4, by my understanding you'd need to encode the end coord as 2.5 to achieve the desired effect. But you can't get 2.5 into an F2Dot14.

So given that the instancer has to be able to do this, I wonder if it shouldn't go ahead and do it in all cases in order to remain compliant with the current expectations of the spec. Given the explicit spec requirement you quoted above, I'm slightly concerned that some implementations might not handle values outside of the [-1.0, 1.0] range correctly.

**Comment by anthrotype** [comment]
Of course the instancer can and does create additional tuples when the coords don't fit the max F2Dot14 limits. 
The reason we would like to avoid in the case when they do fit is that it saves bytes.
Implementation do not need to handle anything outside the -1.0/+1.0 limits, they should continue to clamp and assume 1.0 is the maximum and -1.0 the minimum.
They just have to be able to decode F2Dot14 tuple coordinates, which of course all can, and don't outright reject it if it's greater than 1.0 (or less than -1.0).
I don't even know why that limitation was explicitly written like that on the spec.
Do we know of any implementations (besides OTS) that isn't happy with the tuples we emit?
